---
title: "Comparing the transcriptional response to Heat Shock in embryos and adults"
author: "Thomas O'Leary"
output:
  rmarkdown::html_document:
    theme: lumen
    number_sections: true
    toc: true
    toc_float: true
---

```{r setup, include = FALSE}
# Set options
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

# Load packages
require(tidyverse)
require(kableExtra)

# Data wrangling ----- 

# Load transcript and genes
genes_df <- read_csv(here::here("emily/results/emily_transcript_genes.csv"))

# Load data from Fronteirs
df_f <- read_csv(here::here("cahan/results/Dm_cahan_deg_hot_results_with_norm.csv"))

df_32 <- read_csv(here::here("emily/results/temp_32_25_results_with_norm.csv")) %>%
  rename(transcript_id = gene) %>%
  left_join(genes_df)
df_34 <- read_csv(here::here("emily/results/temp_34_25_results_with_norm.csv")) %>%
  rename(transcript_id = gene) %>%
  left_join(genes_df)
df_36 <- read_csv(here::here("emily/results/temp_36_25_results_with_norm.csv")) %>%
  rename(transcript_id = gene) %>%
  left_join(genes_df)

# Row bind the dataframe
df <- bind_rows(list("adult" = df_f, 
                     "32" = df_32, 
                     "34" = df_34, 
                     "36" = df_36), 
                     .id = "heat")
```

**Quick Disclaimers**: 

- Not a perfect comparision. 

- We don't have adult transcriptional data for the same natural lines. So I am taking the adult HSR data from Canton-S flies. RNA was extracted from whole-bodies after heat shock.

- The 37°C heat shock of the adults is a sub-lethal heat shock. But the heat shock of the embryos (34°C & 36°C) are much closer to their thermal limit (right around their LT<sub>50</sub>).

- The embryo data was analyzed at the transcript level, but the adult data was at the gene level. I have quickly condensed the embryo data by looking only at the most abundant transcript isoform (to characterize whether that gene was up- or down-regulated). 

# Superoxide dismutase

Superoxide dismutase is heat inducible in adult flies.

## Adult (Canton-S) 37°C vs ctrl

```{r}
df_f %>%
  filter(str_detect(gene, "Sod") == TRUE) %>%
  filter(str_detect(gene, "Sodh") == FALSE) %>%
  select(gene, log2FoldChange, baseMean, padj) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

## Embryo (natural lines) HS v ctrl

```{r}
df %>%
  filter(str_detect(gene, "Sod") == TRUE) %>%
  filter(str_detect(gene, "Sodh") == FALSE) %>%
  filter(heat != "adult") %>%
  select(gene, transcript_id, heat, log2FoldChange, baseMean, padj) %>%
  arrange(padj) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(height = "500px")
```

The data I showed in the presentation for FBtr0089938 was thermally responsive for the tropical embryos and not the temparate embryos. There is no significant differential expression of any superoxide dismutase when the natural lines from temperate and tropical regions are grouped together (as they are in this data).

# Overall Heat Shock Response

## Number of expressed genes

```{r}
# Total number of genes or transcripts -----------------------------------------

# # Count the number of genes or transcripts expressed!
# df %>%
#   group_by(heat) %>%
#   filter(!is.na(padj)) %>%
#   count() %>%
#   kable() %>%
#   kable_styling("striped") %>%
#   scroll_box(width = "100%", height = "100%")

# Count the number of unique genes expressed
df %>%
  group_by(heat, gene) %>%
  top_n(n = 1, wt = baseMean) %>%
  filter(!is.na(padj)) %>%
  ungroup(gene) %>%
  count() %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")

```

The embryos express less genes overall than the adults. Approximately 40% less overall number of genes expressed in the emrbyos.


## Number of DEGs

```{r}
# Total number of significantly up or down DEG(/Transcripts) in each -----------

# Filter only significant genes & transcripts
df_sig <- df %>%
  filter(padj < 0.05) 

df %>%
  filter(!is.na(padj)) %>%
  group_by(heat, padj < 0.05) %>%
  count() %>%
  group_by(heat) %>% 
  mutate(per_total = n/sum(n)*100) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

A smaller fraction of the expressed genes are differentially expressed under heat shock in embryos compared to adults.

```{r}
# Count unique genes significantly up or down in each
df_sig %>%
  group_by(heat, log2FoldChange < 0, gene) %>%
  top_n(n = 1, wt = baseMean) %>%
  ungroup(gene) %>%
  count() %>%
  group_by(heat) %>% 
  mutate(per_total = n/sum(n)*100) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

Of the genes that are significantly differentially expressed, only ~ 28% are up-regulated in the adult, where in the embryo ~ 48% of its genes are up-regulated.

## Number of shared DEGs

### Up-regulated

```{r}

# Overlap between temp responsive genes in each category -----------------------

# Up-regulated genes
df_sig %>%
  group_by(heat, gene) %>%
  top_n(n = 1, wt = baseMean) %>%
  filter(log2FoldChange > 0) %>%
  count() %>%
  spread(heat, n, fill = 0) %>%
  ungroup(gene) %>%
  select(-gene) %>% 
  as.matrix() %>% 
  crossprod() %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

There are a decent amount of genes that are thermally responsive only in the embryos and not the adults. ~ 62% of the up-regulated genes in the 36°C heat shock.

### Down-regulated

```{r}
# Down-regulated genes
df_sig %>%
  group_by(heat, gene) %>%
  top_n(n = 1, wt = baseMean) %>%
  filter(log2FoldChange < 0) %>%
  count() %>%
  spread(heat, n, fill = 0) %>%
  ungroup(gene) %>%
  select(-gene) %>% 
  as.matrix() %>% 
  crossprod() %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```


```{r}
# Shared up-regulated gene list ------------------------------------------------

# Which genes are shared?
z <- df_sig %>%
  group_by(heat, gene)  %>%
  top_n(n = 1, wt = baseMean) %>%
  filter(log2FoldChange > 0) %>%
  group_by(gene) %>%
  mutate(n = n()) %>%
  filter(n >= 2) %>%
  select(gene, heat) %>%
  group_by(gene) %>%
  summarise(groups = paste(sort(unique(heat)), collapse = ",")) %>%
  arrange(groups)

z_shr <- z %>%
  filter(str_detect(groups, "36,adult") == TRUE)

uni_36 <- df_36 %>%
  filter(padj < 0.05) %>%
  filter(log2FoldChange > 0) %>%
  filter(!(gene %in% z_shr$gene))

uni_adult <- df_f %>%
  filter(padj < 0.05) %>%
  filter(log2FoldChange > 0) %>%
  filter(!(gene %in% z_shr$gene))

write_tsv(as.data.frame(unique(z_shr$gene)),
          here::here("emily/results/shared_embryo_36_adult_up_genes.txt"))

write_tsv(as.data.frame(unique(uni_36$gene)),
          here::here("emily/results/unique_embryo_36_up_genes.txt"))

write_tsv(as.data.frame(unique(uni_adult$gene)),
          here::here("emily/results/unique_adult_up_genes.txt"))
```
