---
title: Sketch of the final figures
output:
  rmarkdown::html_document:
    theme: lumen
    number_sections: false
    toc: true
    toc_float: true
# output:
#   pdf_document:
#     latex_engine: xelatex
# mainfont: Arial
# fontsize: 14pt
# geometry: margin=2cm
# headerincludes:
#   - \usepackage{float}
#   - \usepackage[fontsize=20pt]{scrextend}
---

```{r setup, include=FALSE}
# don't show cose, warnings or messages
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
# don't make the log files for the VennDiagram package
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

# load these required packages
require(tidyverse)
require(ggpubr)
require(seqsetvis)
require(ssvRecipes)
require(knitr)
require(kableExtra)

# source the functions and load the data
source(here::here("functions.R"))
source(here::here("cahan/scripts/load_data.R"))
```

# **Figure 1:** Phenotypes

```{r, fig.height = 5, fig.width = 7}
p1 <- ggplot(pheno %>%
               filter(type == "ctmin") %>%
               group_by(line_DGRP) %>%
               summarize(temp = mean(temp)), 
             aes(x = temp)) + 
        geom_histogram(binwidth = 0.33, color = "grey", alpha = 0.75) + 
        labs(x = expression(CT[min] * ~ "(째C)"),
             y = "Frequency") +
        xlim(c(0, 10)) +
        ylim(c(0, 33)) + 
        annotate("text", x = 7.5, y = 17.5, label = expression(H^2 *" = 0.25"),
                 size = 4) +
        theme_classic()

p2 <- ggplot(pheno %>%
               filter(type == "ctmax") %>%
               group_by(line_DGRP) %>%
               summarize(temp = mean(temp)), 
             aes(x = temp)) + 
        geom_histogram(binwidth = 0.33, color = "grey", alpha = 0.75) + 
        labs(x = expression(CT[max] * ~ "(째C)"), 
             y = "Frequency") +
        xlim(c(35, 45)) +
        ylim(c(0, 33)) +
        annotate("text", x = 42.5, y = 17.5, label = expression(H^2 *" = 0.29"),
                 size = 4) +
        theme_classic()

top_row <- cowplot::plot_grid(p1, p2, nrow = 1)

my.formula <- y ~ x
df <- pheno %>%
        pivot_wider(names_from = type, values_from = temp) %>%
        group_by(line_DGRP) %>%
        summarize(ctmin = mean(ctmin), ctmax = mean(ctmax))

p3 <- ggplot(df, 
             aes(x = ctmin, y = ctmax)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = "right", label.y = "bottom") + 
        annotate("text", x = 3, y = 40.5, 
                 label = paste("r =", round(cor(df$ctmin, df$ctmax), 2)), 
                 size = 4) +
        labs(x = expression(CT[min] * ~ "(째C)"), 
             y = expression(CT[max] * ~ "(째C)")) +
        # xlim(c(0, 10)) +
        # ylim(c(38.5, 41)) +
        theme_classic() +
        theme(plot.margin = unit(c(t = 0.25, r = 3.5, b = 0, l = 3.5), "cm"))

cowplot::plot_grid(top_row, p3, nrow = 2)

setwd(here::here("cahan/plots"))
ggsave("fig_1_final.tiff", 
       units = "in", 
       width = 7, 
       height = 5, 
       dpi=300, 
       compression = 'lzw')
```

**Figure 1: Distribution of thermal limits in the DGRP.** Histograms of mean phenotypes across 100 DGRP lines for **(A)** CT<sub>min</sub> and **(B)** CT<sub>max</sub>. **(C)** Correlation between line-means of CT<sub>min</sub> and CT<sub>max</sub>.




# **Figure 2:** Manhattan plots

```{r, fig.height = 5, fig.width = 7}

p1 <- cowplot::ggdraw() +
  cowplot::draw_image(here::here("DRGP_GWAS/gwas_cold_manhattan.png"), scale = 1) +
  cowplot::draw_label(expression(CT[min]), y = 0.9, size = 18, fontface = "bold")

p2 <- cowplot::ggdraw() +
  cowplot::draw_image("/Users/tsoleary/R/rna_seq/cahan/plots/gwas_4_venn_diagram_ct.png", 
                      scale = 1) +
  cowplot::draw_label("Allelic variants", y = 0.95, size = 18)#, fontface = "bold")

p_top <- cowplot::plot_grid(p1, p2, nrow = 1, rel_widths = c(2,1))


p1 <- cowplot::ggdraw() + 
  cowplot::draw_image(here::here("DRGP_GWAS/gwas_hot_manhattan.png"), scale = 1) +
  cowplot::draw_label(expression(CT[max]), y = 0.9, size = 18, fontface = "bold")

p2 <- cowplot::ggdraw() +
  cowplot::draw_image("/Users/tsoleary/R/rna_seq/cahan/plots/gwas_4_genes_venn_diagram_ct.png", scale = 1) +
  cowplot::draw_label("Genes", y = 0.95, size = 18)#, fontface = "bold")

p_bottom <- cowplot::plot_grid(p1, p2, nrow = 1, rel_widths = c(2,1))

cowplot::plot_grid(p_top, p_bottom, nrow = 2)

setwd(here::here("cahan/plots"))
ggsave("fig_2_final.tiff", 
       units = "in", 
       width = 7, 
       height = 5, 
       dpi=300, 
       compression = 'lzw')
```

**Figure 2: Results of GWAS to identify polymorphisms associated with thermal tolerance.** Manhattan plots of the results from the GWAS for **(A)** CT<sub>min</sub> and **(B)** CT<sub>max</sub>. The blue line corresponds to p < 1E-4 and the red line corresponds to p < 1E-5. The overlap of **(C)** allelic variants and **(D)** unique genes between CT<sub>min</sub> and CT<sub>max</sub> is also shown.


# **Figure 3:** Effect size & MAF

```{r, fig.height = 6.5, fig.width = 5.25}

x <- pheno %>%
  group_by(line_DGRP, type) %>%
  summarize(temp = mean(temp)) %>%
  ungroup(line_DGRP) %>%
  group_by(type) %>%
  summarize(sd = sd(temp))

MAF_df <- bind_rows(
  gwas_hot_5 %>%
    mutate(NormEff = AvgEff / x$sd[x$type == "ctmax"],
           GWAS = "CTmax") %>%
    select(NormEff, MAF, GWAS),
  gwas_cold_5 %>%
    mutate(NormEff = (AvgEff / x$sd[x$type == "ctmin"]) * -1,
           GWAS = "CTmin") %>%
    select(NormEff, MAF, GWAS))


p1 <- ggplot(MAF_df, 
  aes(x = NormEff, y = MAF, fill = GWAS, shape = GWAS)) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0) +
  labs(x = expression("Tolerance effect size  x  " * sigma[p]^-1), 
       y = "Minor allele frequency") +
  xlim(c(-1, 1)) +
  ylim(c(0, 0.5)) +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank(),
        legend.position = c(0.9, 0.9)) +
  scale_shape_manual(values = c(24, 21),
                     breaks = c("CTmax", "CTmin"),
                     labels = c(expression(CT[max]), expression(CT[min]))) +
  scale_fill_manual(values = c("indianred1", "lightskyblue"),
                    breaks = c("CTmax", "CTmin"),
                    labels = c(expression(CT[max]), expression(CT[min])))

p2 <- ggplot(MAF_df,
  aes(x = NormEff, fill = GWAS)) +
  stat_bin(color = "grey50", position = "dodge", bins = 30) +
  geom_vline(xintercept = 0) +
  labs(x = expression("Tolerance effect size  x  " * sigma[p]^-1),
       y = "Frequency") +
  xlim(c(-1,1)) +
  theme_classic(base_size = 14) +
  scale_fill_manual(values = c("indianred1", "lightskyblue"),
                    breaks = c("CTmax", "CTmin"),
                    labels = c(expression(CT[max]), expression(CT[min]))) +
  theme(legend.title = element_blank(),
        legend.position = c(0.9, 0.9))




# # original split graphs...
# p1 <- ggplot(gwas_hot_5 %>%
#                mutate(NormEff = AvgEff / x$sd[x$type == "ctmax"]), 
#         aes(x = NormEff, y = MAF)) +
#         geom_point(fill = "indianred1", shape = 21) +
#         geom_vline(xintercept = 0) +
#         labs(x = expression("Tolerance effect size  x  " * sigma[p]^-1), 
#              y = "Minor allele frequency") +
#         xlim(c(-1,1)) +
#         ylim(c(0,0.5)) +
#         theme_classic()
# 
# p2 <- ggplot(gwas_cold_5 %>%
#                mutate(NormEff = (AvgEff / x$sd[x$type == "ctmin"]) * -1), 
#         aes(x = NormEff, y = MAF)) +
#         geom_point(fill = "lightskyblue", shape = 21) +
#         geom_vline(xintercept = 0) +
#         labs(x = expression("Tolerance effect size  x  " * sigma[p]^-1), 
#              y = "Minor allele frequency") +
#         xlim(c(-1,1)) +
#         ylim(c(0,0.5)) +
#         theme_classic()
#
cowplot::plot_grid(p2, p1, rel_heights = c(1.625, 2), nrow = 2)

setwd(here::here("cahan/plots"))
ggsave("fig_3_final.tiff", 
       units = "in", 
       width = 5.25, 
       height = 6.5, 
       dpi=300, 
       compression = 'lzw')
```

**Figure 3:** Scatter plot of the minor allele frequency versus the tolerance effect size scaled by the phenotypic standard deviation ($\sigma_{p}$) for CT<sub>min</sub> and CT<sub>max</sub>. 


Positive values correspond to increased cold tolerance or increased heat tolerance.



# **Figure 4:** DEG & GWAS integration

```{r, fig.height = 7, fig.width = 7}
# filter the gwas so there is only the pvalues that reach the cut off level ----
pval_cut_4 <- 10^-4

gwas_cold_4 <- gwas_cold %>%
  filter(AvgMixedPval < pval_cut_4)
gwas_hot_4 <- gwas_hot %>%
  filter(AvgMixedPval < pval_cut_4)

# create a data.frame with all the info for the plotting later -----------------
temp <- left_join(deg, gwas_cold_4, by = "gene", suffix = c("", ".cold"))
total_4 <- full_join(temp, gwas_hot_4, by = "gene", suffix = c(".cold", ".hot"))

# group by the significance
deg_grouped_4 <- group_deg(total_4, pval_cut = 0.01, lfc_cut = 0)


color_set <- c("tan4", "tan", "coral", "darkorchid1", "grey50")
color_set <- c("goldenrod4", "goldenrod", "coral", "darkorchid1", "grey50")

# the function from seqsetvis requires a sequential column in the data.frame
deg_grouped_4$id <- seq_len(nrow(deg_grouped_4))

deg_simple_4 <- deg_grouped_4 %>%
  dplyr::select(gene, log2FoldChange.hot, 
                log2FoldChange.cold, groupn, id, ID.hot, ID.cold) %>%
  mutate(gwas_g = case_when(is.na(ID.hot) & is.na(ID.cold) ~ "Other",
                            !is.na(ID.hot) & is.na(ID.cold)  ~ "CTmax",
                            is.na(ID.hot) & !is.na(ID.cold)  ~ "CTmin",
                            !is.na(ID.hot) & !is.na(ID.cold)  ~ "CTmin_CTmax"))

deg_simple_distinct_4 <- deg_simple_4 %>%
  distinct(gene, .keep_all = TRUE)


deg_simple_distinct_4 <- deg_simple_distinct_4 %>%
  separate(gwas_g, into = c("gwas_g", "gwas_g2"), sep = "_") %>%
  pivot_longer(gwas_g:gwas_g2, names_to = "type", values_to = "gwas_g", 
               values_drop_na = TRUE)

# order the levels of the factor for plotting purposes to get NS on bottom
deg_simple_distinct_4$groupn <- as.factor(deg_simple_distinct_4$groupn)

deg_simple_distinct_4$groupn <- factor(deg_simple_distinct_4$groupn, 
                                       levels =  c("Shared: 6228",
                                                   "Sig1 Shared: 3266", 
                                                   "Unique: 14",
                                                   "Sig1 Unique: 106",
                                                   "NS: 6485"))


plot_scatter_side_density_xy_rel(deg_simple_distinct_4, 
                             x_ = "log2FoldChange.cold", 
                             y_ = "log2FoldChange.hot",
                             labs_x = "Cold vs Ctrl (Log2 fold-change)",
                             labs_y = "Hot vs Ctrl (Log2 fold-change)",
                             bg.string = "NS: 6485",
                             bg.density.string = "Other",
                             labs_sets_density = "Top GWAS-associated genes",
                             id_ = "id", 
                             set_ = "groupn", 
                             labs_sets = "Differentially expressed\ngenes",
                             set_density_ = "gwas_g", 
                             sets.density.colors = c("red", "blue", "yellow"),
                             sets.colors = color_set,
                             n_auto_label = 0)

setwd(here::here("cahan/plots"))
ggsave("fig_4_final_colors2.tiff", 
       units = "in", 
       width = 7, 
       height = 7, 
       dpi=300, 
       compression = 'lzw')
```


**Figure 4: Scatter plot of the gene expression log-fold change of heat-shock (37째C) and cold-shock (4째C) relative to the 25째C control.** The density plots indicate the density of the genes that are associated with CT<sub>min</sub> (blue) or CT<sub>max</sub> (red) relative to the log fold change expression of all other genes (gray). 




# **Figure 5:** ORA for GWAS

![](https://tsoleary.github.io/rna_seq/cahan/results/GO_results_files/go_fig_lfc_order.png)

**Figure 5:  Expression patterns of GWAS genes within significantly over-represented GO biological process categories.** All expression patterns are expressed relative to the 25째C control. Bolded categories indicate those also over-represented in the full DEG dataset for the corresponding temperature extreme. Bolded genes indicate significant differential gene expression at an FDR < 0.01.

# **Table 1:** Gene counts

```{r}
comb_cold <- left_join(gwas_cold, res_cold, by = "gene")
comb_hot <- left_join(gwas_hot, res_hot, by = "gene")

# create a column that indicates where the snps are from
comb_hot$gwas <- "CTmax"
comb_cold$gwas <- "CTmin"

comb_gwas <- dplyr::bind_rows(comb_hot, comb_cold)


t1 <- comb_gwas %>%
  group_by(gwas) %>%
  distinct(ID, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("Relaxed SNPs" = n)

t2 <- comb_gwas %>%
  filter(AvgMixedPval < 10^-5) %>%
  group_by(gwas) %>%
  distinct(ID, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("Strict SNPs" = n)

t3 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("Unique genes" = n)

t4 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::filter(padj <= 0.01) %>%
  dplyr::count() %>%
  dplyr::rename("DEGs" = n)

t5 <- comb_gwas %>%
  dplyr::group_by(gwas) %>%
  dplyr::filter(gene %in% dmel_tf$SYMBOL) %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("TFs" = n)

t6 <- comb_gwas %>%
  dplyr::group_by(gwas) %>%
  dplyr::filter(gene %in% dmel_tf$SYMBOL) %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::filter(padj <= 0.01) %>%
  dplyr::count() %>%
  dplyr::rename("DE TFs" = n)

table <- full_join(t1, t2) %>%
  full_join(t3) %>%
  full_join(t4) %>%
  full_join(t5) %>%
  full_join(t6) %>%
  rename(GWAS = gwas)

table %>%
  kable() %>%
  kable_styling("striped") %>%
  row_spec(0, bold = TRUE, font_size = 12)
```


**Table 1: Counts of the  CTmax and CTmin SNPs and integrated expression data.** Counts of Relaxed SNPs (p < 10$^{-4}$), Strict SNPs (p < 10$^{-5}$), unique genes, differentially expressed genes (DEGs; FDR < 0.01), transcription factors (TFs), and differentially expressed transcription factors (DE TFs).

# **Table 2:** Feature counts

```{r}
comb_gwas %>%
  group_by(gwas, feature) %>%
  dplyr::count() %>%
  filter(feature != "") %>%
  arrange(desc(n)) %>%
  pivot_wider(names_from = "gwas", values_from = "n") %>%
  ungroup(feature) %>%
  mutate(CTmax = replace_na(CTmax, 0),
         feature = str_replace_all(str_to_sentence(feature), "_", " ")) %>%
  mutate(feature = case_when(feature == "Synonymous coding" ~ "Synonymous coding",
                             feature == "Intron" ~ "Intron",
                             feature == "Downstream" ~ "Downstream",
                             feature == "Upstream" ~ "Upstream",
                             feature == "Utr 3 prime" ~ "3' UTR",
                             feature == "Utr 5 prime" ~ "5' UTR",
                             feature == "Non synonymous coding" ~ "Non-synonymous coding",
                             feature == "Exon" ~ "Exon",
                             feature == "Start gained" ~ "Start gained")) %>%
  rename(Feature = feature) %>%
  kable() %>%
  kable_styling("striped") %>%
  row_spec(0, bold = TRUE, font_size = 12)
```

**Table 2: Counts of number of SNPs associated with CTmin and CTmax (p < 10-4) that lie within each genomic feature group.**


# **Suplementary Information**

## Thermal Breadth Correlations

```{r, fig.height = 6, fig.width = 7}
p1 <- ggplot(pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         group_by(line_DGRP) %>%
         summarize(tb = mean(tb)), 
             aes(x = tb)) + 
        geom_histogram(binwidth = 0.33, color = "grey", alpha = 0.75) + 
        labs(x ="Thermal Breadth (째C)",
             y = "Frequency") +
        xlim(c(30, 40)) +
        theme_classic()


df2 <- pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         select(line_DGRP, sex, tb) %>%
         pivot_wider(names_from = sex, values_from = tb)

my.formula <- y ~ x
p2 <- ggplot(df2, 
        aes(x = male, y = female)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        annotate("text", x = 32.25, y = 37.5, 
                 label = paste("r =", round(cor(df2$male, df2$female), 2)), 
                 size = 4) +
        labs(x = "Male Thermal Breadth (째C)", 
             y = "Female Thermal Breadth (째C)") +
        theme_classic()

df3 <- pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         group_by(line_DGRP) %>%
         summarize(tb = mean(tb), ctmin = mean(ctmin), ctmax = mean(ctmax))

my.formula <- y ~ x
p3 <- ggplot(df3, 
        aes(x = ctmin, y = tb)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        annotate("text", x = 3, 33,
                 label = paste("r =", round(cor(df3$ctmin, df3$tb), 2)),
                 size = 4) +
        labs(x = expression(CT[min] * ~ "(째C)"), 
             y = "Thermal Breadth (째C)") +
        theme_classic()
  

p4 <- ggplot(pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         group_by(line_DGRP) %>%
         summarize(tb = mean(tb), ctmin = mean(ctmin), ctmax = mean(ctmax)), 
        aes(x = ctmax, y = tb)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        annotate("text", x = 39, y = 38, 
                 label = paste("r =", round(cor(df3$ctmax, df3$tb), 2)), 
                 size = 4) +
        labs(x = expression(CT[max] * ~ "(째C)"), 
             y = "Thermal Breadth (째C)") +
        theme_classic()

cowplot::plot_grid(p1, p2, p3, p4, nrow = 2)


setwd(here::here("cahan/plots"))
ggsave("sup_fig_1_final.tiff", 
       units = "in", 
       width = 6, 
       height = 7, 
       dpi=300, 
       compression = 'lzw')
```

**Supplementary Figure X: (A)** Histograms of mean phenotypes across 100 DGRP lines for thermal breadth. **(B)** Correlation between sex-means for thermal breadth. Correlation between line-means of **(C)** thermal breadth and CT<sub>min</sub> and **(D)** thermal breadth and CT<sub>max</sub>.



```{r, fig.height = 3, fig.width = 7}
df <- pheno %>%
        pivot_wider(names_from = sex, values_from = temp) %>%
        pivot_wider(names_from = type, values_from = c("male", "female"))

p1 <- ggplot(df, 
        aes(x = male_ctmin, y = female_ctmin)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        annotate("text", x = 2, y = 8,
                 label = paste("r =",
                               round(cor(df$male_ctmin, df$female_ctmin), 2)),
                 size = 4) +
        labs(x = expression("Male" ~ CT[min] * ~ "(째C)"), 
             y = expression("Female" ~ CT[min] * ~ "(째C)")) +
        theme_classic()

p2 <- ggplot(df, 
        aes(x = male_ctmax, y = female_ctmax)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        # ggpmisc::stat_poly_eq(formula = my.formula, 
        #                       aes(label = ..rr.label..),
        #                       parse = TRUE, label.x = 0.1, label.y = 0.95) +
        annotate("text", x = 39, y = 40.5,
                 label = paste("r =",
                               round(cor(df$male_ctmax, df$female_ctmax), 2)),
                 size = 4) +
        labs(x = expression("Male" ~ CT[max] * ~ "(째C)"), 
             y = expression("Female" ~ CT[max] * ~ "(째C)")) +
        theme_classic()

cowplot::plot_grid(p1, p2, nrow = 1)

setwd(here::here("cahan/plots"))
ggsave("sup_fig_2_final.tiff", 
       units = "in", 
       width = 7, 
       height = 3,
       dpi=300, 
       compression = 'lzw')

```

**Supplementary Figure X:** Correlation between sex-means for **(A)** CT<sub>min</sub> and **(B)** CT<sub>max</sub>.

## Links to download supplementary tables

[ORA GWAS & DEG Tables](https://tsoleary.github.io/rna_seq/cahan/results/GO_results_files/ora_gwas_deg_suppl.xlsx)

[DEG Results & Tables](https://tsoleary.github.io/rna_seq/cahan/results/deg_suppl_info.xlsx)

