---
title: Sketch of the final figures
output:
  rmarkdown::html_document:
    theme: lumen
    number_sections: false
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
# don't show cose, warnings or messages
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
# don't make the log files for the VennDiagram package
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

# load these required packages
require(tidyverse)
require(ggpubr)
require(AnnotationDbi)
require(org.Dm.eg.db)
require(ggcorrplot)
require(seqsetvis)
require(ssvRecipes)
require(knitr)
require(kableExtra)
source(here::here("functions.R"))

# load the differentially expressed genes results
setwd(here::here("cahan/results"))
res_cold <- read.csv(list.files(pattern = "cold_results"))
res_hot <- read.csv(list.files(pattern = "hot_results"))
deg <- full_join(res_cold, res_hot, by = "gene", suffix = c(".cold", ".hot"))

# load the gwas results
setwd(here::here("DRGP_GWAS"))
pheno <- read.csv("pheno.csv")
gwas_cold_all <- read.table("CTmin/gwas.top.4.annot", header = TRUE)
gwas_hot_all <- read.table("CTmax/gwas.top.4.annot", header = TRUE)
gwas_tb_all <- read.table("ThermalBreadth/gwas.top.4.annot", header = TRUE)

# # load all the gwas snps for the manhattan plot
# gwas_cold_manhat <- read_delim("CTmin/gwas.all.assoc", 
#                                delim = " ", 
#                                col_names = TRUE)
# gwas_hot_manhat <- read_delim("CTmax/gwas.all.assoc", 
#                               delim = " ", 
#                               col_names = TRUE)
# 
# # wrangle the manhat dfs for qqman
# gwas_cold_manhat <- gwas_cold_manhat %>%
#   separate(ID, into = c("CHR", "BP", "type")) %>%
#   mutate(CHR = case_when(CHR == "2L" ~ 1,
#                          CHR == "2R" ~ 2,
#                          CHR == "3L" ~ 3,
#                          CHR == "3R" ~ 4,
#                          CHR == "4" ~ 5,
#                          CHR == "X" ~ 6)) %>%
#   dplyr::filter(!is.na(AvgMixedPval) | !is.na(CHR) | !is.na(BP))
# 
# gwas_cold_manhat$P <- gwas_cold_manhat$AvgMixedPval
# gwas_cold_manhat$BP <- as.numeric(gwas_cold_manhat$BP)
# 
# 
# gwas_hot_manhat <- gwas_hot_manhat %>%
#   separate(ID, into = c("CHR", "BP", "type")) %>%
#   mutate(CHR = case_when(CHR == "2L" ~ 1,
#                          CHR == "2R" ~ 2,
#                          CHR == "3L" ~ 3,
#                          CHR == "3R" ~ 4,
#                          CHR == "4" ~ 5,
#                          CHR == "X" ~ 6)) %>%
#   dplyr::filter(!is.na(AvgMixedPval) | !is.na(CHR) | !is.na(BP))
# 
# gwas_hot_manhat$P <- gwas_hot_manhat$AvgMixedPval
# gwas_hot_manhat$BP <- as.numeric(gwas_hot_manhat$BP)


# wrangle the phenotypes
pheno <- pheno %>%
  pivot_longer(cols = starts_with("ct"), 
               names_to = "type",
               values_to = "temp") %>%
  separate(type, into = c("type", "sex") ,sep = "_")
  

# annotated all the gene symbols for the gwas output  --------------------------

gwas_cold_all <- gwas_dgrp_gene_annot(gwas_cold_all)
gwas_hot_all <- gwas_dgrp_gene_annot(gwas_hot_all)
gwas_tb_all <- gwas_dgrp_gene_annot(gwas_tb_all)

# filter the gwas so there is only the pvalues that reach the cut off level ----
pval_cut <- 10^-5

gwas_cold <- gwas_cold_all %>%
  filter(AvgMixedPval < pval_cut)
gwas_hot <- gwas_hot_all %>%
  filter(AvgMixedPval < pval_cut)
gwas_tb <- gwas_tb_all %>%
  filter(AvgMixedPval < pval_cut)

# create a combined df with deg and gwas for each alone ------------------------
comb_cold <- full_join(res_cold, gwas_cold, by = "gene")
comb_hot <- full_join(res_hot, gwas_hot, by = "gene")
  
# create a data.frame with info for the individual ma_plots --------------------
deg_gwas_cold <- left_join(res_cold, gwas_cold, by = "gene")
deg_gwas_hot <- left_join(res_hot, gwas_hot, by = "gene")

# create a data.frame with all the info for the plotting later -----------------
temp <- left_join(deg, gwas_cold, by = "gene", suffix = c("", ".cold"))
total <- full_join(temp, gwas_hot, by = "gene", suffix = c(".cold", ".hot"))
#total <- full_join(total, gwas_tb, by = "gene", suffix = c("", ".tb"))
# the termal breadth stuff to the 
 
# group by the significance
deg_grouped <- group_deg(total, pval_cut = 0.01, lfc_cut = 0)

# colors for the shared unique scatter plot  
##############
# create a parameters space where all these go later .....
clrs <- c("darkgrey", "goldenrod4", "goldenrod1", "coral", "coral4")
legend.font.size <- 12
```

# Figure 1: Phenotypes

```{r, fig.height = 5, fig.width = 7}
p1 <- ggplot(pheno %>%
               filter(type == "ctmin") %>%
               group_by(line_DGRP) %>%
               summarize(temp = mean(temp)), 
             aes(x = temp)) + 
        geom_histogram(binwidth = 0.33, color = "grey", alpha = 0.75) + 
        labs(x = expression(CT[min] * ~ "(째C)"),
             y = "Counts") +
        xlim(c(0, 10)) +
        ylim(c(0, 33)) + 
        annotate("text", x = 7.5, y = 17.5, label = expression(H^2 *" = 0.25"),
                 size = 4) +
        theme_classic()

p2 <- ggplot(pheno %>%
               filter(type == "ctmax") %>%
               group_by(line_DGRP) %>%
               summarize(temp = mean(temp)), 
             aes(x = temp)) + 
        geom_histogram(binwidth = 0.33, color = "grey", alpha = 0.75) + 
        labs(x = expression(CT[max] * ~ "(째C)"), 
             y = "Counts") +
        xlim(c(35, 45)) +
        ylim(c(0, 33)) +
        annotate("text", x = 42.5, y = 17.5, label = expression(H^2 *" = 0.29"),
                 size = 4) +
        theme_classic()

top_row <- cowplot::plot_grid(p1, p2, nrow = 1, labels = "AUTO")

my.formula <- y ~ x
p3 <- ggplot(pheno %>%
               pivot_wider(names_from = type, values_from = temp) %>%
               group_by(line_DGRP) %>%
               summarize(ctmin = mean(ctmin), ctmax = mean(ctmax)), 
             aes(x = ctmin, y = ctmax)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        ggpmisc::stat_poly_eq(formula = my.formula, 
                              aes(label = ..rr.label..),
                              parse = TRUE, label.x = "right", label.y = "bottom") + 
        labs(x = expression(CT[min] * ~ "(째C)"), 
             y = expression(CT[max] * ~ "(째C)")) +
        theme_classic() +
        theme(plot.margin = unit(c(t = 0.25, r = 3.5, b = 0, l = 3.5), "cm"))

cowplot::plot_grid(top_row, p3, nrow = 2, labels = c("", "C"))
```


# Figure 2: Manhattan plots

```{r, fig.height = 5, fig.width = 7}
# setwd(here::here("DRGP_GWAS"))
# 
# png("gwas_cold_manhattan.png", width = 600, height = 300)
# qqman::manhattan(gwas_cold_manhat,
#                  annotateTop = FALSE,
#                  chrlabs = c("2L", "2R", "3L", "3R", "4", "X"),
#                  ylim = c(0, 8),
#                  suggestiveline = -log10(1e-04),
#                  genomewideline = -log10(1e-05))
# dev.off()
# 
# 
# png("gwas_hot_manhattan.png", width = 600, height = 300)
# qqman::manhattan(gwas_hot_manhat,
#                  annotateTop = FALSE,
#                  chrlabs = c("2L", "2R", "3L", "3R", "4", "X"),
#                  ylim = c(0, 8),
#                  suggestiveline = -log10(1e-04),
#                  genomewideline = -log10(1e-05))
# dev.off()

p1 <- cowplot::ggdraw() +
  cowplot::draw_image(here::here("DRGP_GWAS/gwas_cold_manhattan.png"), scale = 1) +
  cowplot::draw_label("CTmin", y = 0.9, size = 18, fontface = "bold")

p2 <- cowplot::ggdraw() +
  cowplot::draw_image("/Users/tsoleary/R/rna_seq/cahan/plots/gwas_4_venn_diagram_ct.png", 
                      scale = 1) +
  cowplot::draw_label("SNPs", y = 0.95, size = 18, fontface = "bold")

p_top <- cowplot::plot_grid(p1, p2, nrow = 1, rel_widths = c(2,1), labels = c("A", "C"))


p1 <- cowplot::ggdraw() + 
  cowplot::draw_image(here::here("DRGP_GWAS/gwas_hot_manhattan.png"), scale = 1) +
  cowplot::draw_label("CTmax", y = 0.9, size = 18, fontface = "bold")

p2 <- cowplot::ggdraw() +
  cowplot::draw_image("/Users/tsoleary/R/rna_seq/cahan/plots/gwas_4_genes_venn_diagram_ct.png", scale = 1) +
  cowplot::draw_label("Genes", y = 0.95, size = 18, fontface = "bold")

p_bottom <- cowplot::plot_grid(p1, p2, nrow = 1, rel_widths = c(2,1), labels = c("B", "D"))

cowplot::plot_grid(p_top, p_bottom, nrow = 2, labels = c("", ""))
```

# Figure 3: DEG & GWAS integration

```{r, fig.height = 7, fig.width = 7}
# filter the gwas so there is only the pvalues that reach the cut off level ----
pval_cut_4 <- 10^-4

gwas_cold_4 <- gwas_cold_all %>%
  filter(AvgMixedPval < pval_cut_4)
gwas_hot_4 <- gwas_hot_all %>%
  filter(AvgMixedPval < pval_cut_4)

# create a data.frame with all the info for the plotting later -----------------
temp <- left_join(deg, gwas_cold_4, by = "gene", suffix = c("", ".cold"))
total_4 <- full_join(temp, gwas_hot_4, by = "gene", suffix = c(".cold", ".hot"))

# group by the significance
deg_grouped_4 <- group_deg(total_4, pval_cut = 0.01, lfc_cut = 0)


color_set <- c("goldenrod4", "goldenrod1", "coral4", "coral1", "grey50")

# the function from seqsetvis requires a sequential column in the data.frame
deg_grouped_4$id <- seq_len(nrow(deg_grouped_4))

deg_simple_4 <- deg_grouped_4 %>%
  dplyr::select(gene, log2FoldChange.hot, 
                log2FoldChange.cold, groupn, id, ID.hot, ID.cold) %>%
  mutate(gwas_g = case_when(is.na(ID.hot) & is.na(ID.cold) ~ "Other",
                            !is.na(ID.hot) & is.na(ID.cold)  ~ "CTmax",
                            is.na(ID.hot) & !is.na(ID.cold)  ~ "CTmin",
                            !is.na(ID.hot) & !is.na(ID.cold)  ~ "CTmin_CTmax"))

deg_simple_distinct_4 <- deg_simple_4 %>%
  distinct(gene, .keep_all = TRUE)


deg_simple_distinct_4 <- deg_simple_distinct_4 %>%
  separate(gwas_g, into = c("gwas_g", "gwas_g2"), sep = "_") %>%
  pivot_longer(gwas_g:gwas_g2, names_to = "type", values_to = "gwas_g", 
               values_drop_na = TRUE)

# order the levels of the factor for plotting purposes to get NS on bottom
deg_simple_distinct_4$groupn <- as.factor(deg_simple_distinct_4$groupn)

deg_simple_distinct_4$groupn <- factor(deg_simple_distinct_4$groupn, 
                                       levels =  c("Shared: 6228",
                                                   "Sig1 Shared: 3266", 
                                                   "Unique: 14",
                                                   "Sig1 Unique: 106",
                                                   "NS: 6485"))


plot_scatter_side_density_xy_rel(deg_simple_distinct_4, 
                             x_ = "log2FoldChange.cold", 
                             y_ = "log2FoldChange.hot",
                             labs_x = "Cold vs Ctrl (Log2 fold-change)",
                             labs_y = "Hot vs Ctrl (Log2 fold-change)",
                             bg.string = "NS: 6485",
                             bg.density.string = "Other",
                             labs_sets_density = "Top GWAS-associated genes",
                             id_ = "id", 
                             set_ = "groupn", 
                             set_density_ = "gwas_g", 
                             sets.density.colors = c("red", "blue", "yellow"),
                             sets.colors = color_set,
                             n_auto_label = 0)
```

# Figure 4: GO BP GWAS DEG overlap

```{r, fig.height = 9, fig.width = 7.5}
p1 <- cowplot::ggdraw() +
  cowplot::draw_image(here::here("cahan/plots/go_hot_genes.png"))

p2 <- cowplot::ggdraw() +
  cowplot::draw_image(here::here("cahan/plots/go_cold_genes.png"), scale = 1)

cowplot::plot_grid(p1, p2, nrow = 2, rel_heights = c(39,29), labels = "AUTO")
```

# Suplementary Information

## Thermal Breadth Correlations

```{r, fig.height = 3.5, fig.width = 7}
my.formula <- y ~ x
p1 <- ggplot(pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         group_by(line_DGRP) %>%
         summarize(tb = mean(tb), ctmin = mean(ctmin), ctmax = mean(ctmax)), 
        aes(x = ctmin, y = tb)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        ggpmisc::stat_poly_eq(formula = my.formula, 
                              aes(label = ..rr.label..),
                              parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        labs(x = expression(CT[min] * ~ "(째C)"), 
             y = "Thermal Breadth (째C)") +
        theme_classic()
  

p2 <- ggplot(pheno %>%
         pivot_wider(names_from = type, values_from = temp) %>%
         mutate(tb = ctmax - ctmin) %>%
         group_by(line_DGRP) %>%
         summarize(tb = mean(tb), ctmin = mean(ctmin), ctmax = mean(ctmax)), 
        aes(x = ctmax, y = tb)) +
        geom_point(fill = "grey50", shape = 21) +
        geom_smooth(method = "lm", 
                    linetype = "dashed", 
                    se = FALSE, formula = my.formula, color = "grey50") +
        ggpmisc::stat_poly_eq(formula = my.formula, 
                              aes(label = ..rr.label..),
                              parse = TRUE, label.x = 0.1, label.y = 0.95) + 
        labs(x = expression(CT[max] * ~ "(째C)"), 
             y = "Thermal Breadth (째C)") +
        theme_classic()

cowplot::plot_grid(p1, p2, nrow = 1, labels = "AUTO")
```
