---
title: Genome wide association study and differential gene expression of _Drosophila melanogaster_ under heat and cold stress
author: "TS O'Leary"
output:
  rmarkdown::html_document:
    theme: lumen
    number_sections: true
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
# don't show cose, warnings or messages
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

# load these required packages
require(tidyverse)
require(ggpubr)
require(AnnotationDbi)
require(org.Dm.eg.db)
require(ggcorrplot)
require(seqsetvis)
require(ssvRecipes)
require(knitr)
require(kableExtra)
source(here::here("functions.R"))

# load the differentially expressed genes results
setwd(here::here("cahan/results"))
res_cold <- read.csv(list.files(pattern = "cold_results"))
res_hot <- read.csv(list.files(pattern = "hot_results"))
deg <- full_join(res_cold, res_hot, by = "gene", suffix = c(".cold", ".hot"))

# load the gwas results
setwd(here::here("DRGP_GWAS"))
gwas_cold <- read.table("CTmin/gwas.top.annot", header = TRUE)
gwas_hot <- read.table("CTmax/gwas.top.annot", header = TRUE)

# filter the gwas so there is only the pvalues that reach the cut off level ----
pval_cut <- 10^-5

gwas_cold <- gwas_cold %>%
  filter(AvgMixedPval < pval_cut)
gwas_hot <- gwas_hot %>%
  filter(AvgMixedPval < pval_cut)

# combine the cold gwas and deg ------------------------------------------------
# use the GeneAnnotation to get the gene symbol for the gwas
annot_split_cold <- str_split(gwas_cold$GeneAnnotation, "\\|")
# the second item is the gene symbol
gwas_cold$gene <- sapply(annot_split_cold, "[[", 2)
# the third item is where the snp is located, intron, syn coding etc.
gwas_cold$grps <- sapply(annot_split_cold, "[[", 3)
# join the two data frames together
comb_cold <- full_join(res_cold, gwas_cold, by = "gene")

# combine the hot gwas and deg -------------------------------------------------
hot_mult_genes <- str_split(gwas_hot$GeneAnnotation, ";")
annot_split_hot_mult <- sapply(hot_mult_genes, "str_split", "\\|")

gene_list <- vector(mode = "list", length = length(annot_split_hot_mult))

for (i in 1:length(annot_split_hot_mult)) {
  
  list_4  <- which(lengths(annot_split_hot_mult[[i]]) == 4)
  list_10 <- which(lengths(annot_split_hot_mult[[i]]) == 10)
  list_13 <- which(lengths(annot_split_hot_mult[[i]]) == 13)
  
  sym_4 <- c()
  sym_10 <- c()
  sym_13 <- c()
  
  if(!is_empty(list_4)){
      for (k in list_4){
        x <- annot_split_hot_mult[[i]][[c(k, 2)]]
        sym_4 <- c(sym_4, x)
      }
  }
  
  if(!is_empty(list_10)){
      for (k in list_10){
        x <- annot_split_hot_mult[[i]][[c(k, 6)]]
        sym_10 <- c(sym_10, x)
      }
  }
  
  if(!is_empty(list_13)){
      for (k in list_13){
        x <- annot_split_hot_mult[[i]][[c(k, 2)]]
        sym_13 <- c(sym_13, x)
      }
  }
  
  all_symbols <- c(sym_4, sym_10, sym_13)
  
  unique_symbols <- unique(all_symbols[all_symbols != ""])
  
  gene_list[[i]] <- unique_symbols
  
}


## there is either 10 13 or 4 length for each list!!!!! ahhhhhhhhhhhhh fickkkasklfdj

# if 10 then 6
# if 13 then 2
# if 4 then 2

##########


# use the GeneAnnotation to get the gene symbol for the gwas
annot_split_hot <- str_split(gwas_hot$GeneAnnotation, "\\|")
# the second item is the gene symbol
gwas_hot$gene <- sapply(annot_split_hot, "[[", 2)
# the third item is where the snp is located, intron, syn coding etc.
gwas_hot$grps <- sapply(annot_split_hot, "[[", 3)
# join the two data frames together
comb_hot <- full_join(res_hot, gwas_hot, by = "gene")

# create a data.frame with info for the individual ma_plots --------------------
deg_gwas_cold <- left_join(res_cold, gwas_cold, by = "gene")
deg_gwas_hot <- left_join(res_hot, gwas_hot, by = "gene")

# create a data.frame with all the info for the plotting later -----------------
temp <- left_join(deg, gwas_cold, by = "gene", suffix = c("", ".cold"))
total <- full_join(temp, gwas_hot, by = "gene", suffix = c(".cold", ".hot"))

deg_grouped <- total %>%
  mutate(cold_genes = case_when(padj.cold < 0.01 ~ "sig",
                                padj.cold >= 0.01 ~ "ns"),
         hot_genes = case_when(padj.hot < 0.01 ~ "sig",
                               padj.hot >= 0.01 ~ "ns"),
         cold_fc = case_when(log2FoldChange.cold < 0 ~ "down",
                             log2FoldChange.cold > 0 ~ "up"),
         hot_fc = case_when(log2FoldChange.hot < 0 ~ "down",
                            log2FoldChange.hot > 0 ~ "up")) %>%
  filter(!is.na(cold_genes) & !is.na(hot_genes) &
           !is.na(cold_fc) & !is.na(hot_fc)) %>%
  mutate(col = paste(cold_genes, hot_genes, cold_fc, hot_fc)) %>%
  mutate(group = case_when(col == "sig sig up up" |
                             col == "sig sig down down" ~ "Shared",
                           col == "sig ns up up" |
                             col == "ns sig up up" |
                             col == "sig ns down down" |
                             col == "ns sig down down" ~ "Sig1 Shared",
                           col == "ns ns down down" |
                             col == "ns ns down up" |
                             col == "ns ns up down" |
                             col == "ns ns up up" ~ "NS",
                           col == "sig sig down up" |
                             col == "sig sig up down" ~ "Unique",
                           col == "ns sig up down" |
                             col == "ns sig down up" |
                             col == "sig ns down up" |
                             col == "sig ns up down" ~ "Sig1 Unique")) %>%
  add_count(group) %>% 
  mutate(groupn = paste0(group, ': ', n)) %>%
  arrange(groupn)

# colors for the shared unique scatter plot            
clrs <- c("darkgrey", "goldenrod4", "goldenrod1", "coral", "coral4")

legend.font.size <- 12

```

# MA plots {.tabset .tabset-fade .tabset-pills}

All MA plots now have an FDR cutoff of 0.01 and no fold-change cutoff.

## No gene labels

**Cold v Ctrl MA plot**

```{r ma_plot_cold_no_lab}
ggmaplot(res_cold, main = expression("4°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_cold$gene),
         legend = "top", top = 0,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size))
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_no_lab}
ggmaplot(res_hot, main = expression("37°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_hot$gene),
         legend = "top", top = 0,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size))
```

## Labeled by padj

**Cold v Ctrl MA plot**

```{r ma_plot_cold_lab_padj}
ggmaplot(res_cold, main = expression("4°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_cold$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size)) 
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_lab_padj}
ggmaplot(res_hot, main = expression("37°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_hot$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size)) 
```

## GWAS points

**Cold v Ctrl MA plot**

```{r ma_plot_cold_gwas_distinct}
deg_gwas_cold_distinct <- deg_gwas_cold %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_cold_distinct, main = expression("4°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_cold_distinct$gene),
              legend = "top", top = 0,
              select.top.method = "gwas",
              font.label = c("bold", 11), label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size))
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_gwas_distinct}
deg_gwas_hot_distinct <- deg_gwas_hot %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_hot_distinct, main = expression("37°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_hot_distinct$gene),
              legend = "top", top = 0,
              select.top.method = "gwas",
              font.label = c("bold", 11), label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size)) 
```


## GWAS labels

**Cold v Ctrl MA plot**

```{r ma_plot_cold_lab_gwas_distinct}
deg_gwas_cold_distinct <- deg_gwas_cold %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_cold_distinct, main = expression("4°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_cold_distinct$gene),
              legend = "top", top = 14,
              select.top.method = "gwas",
              font.label = c("bold", 11), 
              font.label.sig = c(11, "bold", "black"),
              label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size))
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_lab_gwas_distinct}
deg_gwas_hot_distinct <- deg_gwas_hot %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_hot_distinct, main = expression("37°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_hot_distinct$gene),
              legend = "top", top = 10,
              select.top.method = "gwas",
              font.label = c("bold", 11), 
              font.label.sig = c(11, "bold", "black"),
              label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size)) 
```


## GWAS labels colored by DEG sig

**Cold v Ctrl MA plot**

```{r ma_plot_cold_lab_gwas_deg}
deg_gwas_cold_distinct <- deg_gwas_cold %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_cold_distinct, main = expression("4°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_cold_distinct$gene),
              legend = "top", top = 14,
              select.top.method = "gwas",
              font.label = c("bold", 11), 
              font.label.sig = c(11, "bold", "firebrick"),
              label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size))
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_lab_gwas_deg}
deg_gwas_hot_distinct <- deg_gwas_hot %>%
  distinct(gene, .keep_all = TRUE)

ggmaplot_gwas(deg_gwas_hot_distinct, main = expression("37°C" %->% "25°C"),
              fdr = 0.01, fc = 0, size = 1,
              palette = c("#1465AC", "black", "#A6A6A680", "#B31B21"),
              genenames = as.vector(deg_gwas_hot_distinct$gene),
              legend = "top", top = 10,
              select.top.method = "gwas",
              font.label = c("bold", 11), 
              font.label.sig = c(11, "bold", "firebrick"),
              label.rectangle = TRUE,
              font.legend = "bold",
              font.main = "bold",
              ggtheme = ggplot2::theme_minimal()) + 
  theme(legend.text = element_text(size = legend.font.size)) 
```


# Top 50 differentially expressed genes

## Heat shock

```{r table_top_DEG_hot}
deg_grouped %>%
  distinct(gene, .keep_all = TRUE) %>%
  group_by(hot_fc) %>%
  top_n(n = -50, wt = padj.hot) %>%
  dplyr::select(gene, hot_fc, log2FoldChange.hot, padj.hot) %>%
  arrange(hot_fc, padj.hot) %>%
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 4)))) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

## Cold shock

```{r table_top_DEG_cold}
deg_grouped %>%
  group_by(cold_fc) %>%
  distinct(gene, .keep_all = TRUE) %>%
  top_n(n = -50, wt = padj.cold) %>%
  dplyr::select(gene, cold_fc, log2FoldChange.cold, padj.cold) %>%
  arrange(cold_fc, padj.cold) %>%
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 4)))) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

# Shared and unique - LFC scatter {.tabset .tabset-fade .tabset-pills}

## GWAS scatter & density plots

```{r scatter_and_density_plots, fig.height = 7, fig.width = 7}
color_set <- c("goldenrod4", "goldenrod1", "coral4", "coral1", "grey50")

# the function from seqsetvis requires a sequential column in the data.frame
deg_grouped$id <- seq_len(nrow(deg_grouped))

deg_simple <- deg_grouped %>%
  dplyr::select(gene, log2FoldChange.hot, 
                log2FoldChange.cold, groupn, id, ID.hot, ID.cold) %>%
  mutate(gwas_g = case_when(is.na(ID.hot) & is.na(ID.cold) ~ "Other",
                            !is.na(ID.hot) & is.na(ID.cold)  ~ "CTmax",
                            is.na(ID.hot) & !is.na(ID.cold)  ~ "CTmin",
                            !is.na(ID.hot) & !is.na(ID.cold)  ~ "both"))

# order the levels of the factor for plotting purposes to get NS on bottom
deg_simple$groupn <- as.factor(deg_simple$groupn)

deg_simple$groupn <- factor(deg_simple$groupn, levels = c("Shared: 6113", 
                                                          "Sig1 Shared: 3216", 
                                                          "Unique: 14",
                                                          "Sig1 Unique: 103",
                                                          "NS: 6425"))

deg_simple_distinct <- deg_simple %>%
  distinct(gene, .keep_all = TRUE)

plot_scatter_side_density_xy(deg_simple_distinct, 
                             x_ = "log2FoldChange.cold", 
                             y_ = "log2FoldChange.hot",
                             labs_x = "Cold vs Ctrl (L2FC)",
                             labs_y = "Hot vs Ctrl (L2FC)",
                             bg.string = "NS: 6437",
                             bg.density.string = "Other",
                             labs_sets_density = "GWAS",
                             id_ = "id", 
                             set_ = "groupn", 
                             set_density_ = "gwas_g", 
                             sets.density.colors = c("red", "blue", "yellow"),
                             sets.colors = color_set,
                             n_auto_label = 0)
```

## GWAS density plots (distinct)

```{r density_plots_distinct}
ggplot(deg_simple_distinct) +
  geom_line(mapping = aes(x = log2FoldChange.hot, color = gwas_g), 
            data = deg_simple_distinct, stat = "density") +
  geom_line(mapping = aes(x = log2FoldChange.cold, color = gwas_g), 
            linetype = "dashed", data = deg_simple_distinct, stat = "density") +
  scale_color_manual(values = c("red", "blue", "grey50"), drop = FALSE) +
  labs(x = "log2FoldChange vs. Ctrl", 
       color = "GWAS", 
       caption = "Cold dashed line\nHot solid line") +
  theme_classic()
```

## No labels

```{r shared_unique}
ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  theme(legend.position = "top")
```

## GWAS hits labeled

```{r shared_unique_gwas_lab}
ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, !is.na(ID.cold) | 
                                            !is.na(ID.hot)) %>%
                              distinct(gene, .keep_all = TRUE),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With unique labeled

```{r shared_unique_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Unique") %>%
                              top_n(20, wt = baseMean.cold),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With shared labeled

```{r shared_shared_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Shared") %>%
                              top_n(20, wt = baseMean.cold),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With shared labeled (padj.hot)

```{r shared_shared_padj.hot_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Shared") %>%
                              top_n(20, wt = padj.hot),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

# GWAS Tables

## GWAS summary table

```{r gwas_hot_summary}

comb_hot_filt <- comb_hot %>%
  filter(!is.na(ID))

comb_hot_filt$gwas <- "CTmax"


comb_cold_filt <- comb_cold %>%
  filter(!is.na(ID))

comb_cold_filt$gwas <- "CTmin"

comb_gwas <- rbind(comb_hot_filt, comb_cold_filt)


t1 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::count() %>%
  dplyr::rename("SNPs" = n)

t2 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::count() %>%
  dplyr::rename("SNPs annotated to genes" = n)

t3 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("Unique genes" = n)

t4 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::filter(padj <= 0.01) %>%
  dplyr::count() %>%
  dplyr::rename("Sig DEGs" = n)

t5 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "") %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::filter(padj > 0.01) %>%
  dplyr::count() %>%
  dplyr::rename("NS DEGs" = n)

t6 <- comb_gwas %>%
  group_by(gwas) %>%
  dplyr::filter(gene != "" & is.na(baseMean)) %>%
  dplyr::distinct(gene, .keep_all = TRUE) %>%
  dplyr::count() %>%
  dplyr::rename("No RNAseq data" = n)

t7 <- full_join(t1, t2) %>%
  full_join(t3) %>%
  full_join(t4) %>%
  full_join(t5) %>%
  full_join(t6)

t7$`No RNAseq data` <- t7$`No RNAseq data` %>% replace_na(0)

table <- t7 %>%
  dplyr::rename(GWAS = gwas) 

table %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

## CTmax

### All hits (pval cutoff 10^-5)

```{r table_top_gwas_hot_all}
gwas_hot %>%
  dplyr::select(ID, gene, grps, AvgEff, AvgMixedPval) %>%
  arrange(AvgMixedPval) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

### Group count

```{r table_top_gwas_hot_grps}
gwas_hot %>%
  group_by(grps) %>%
  dplyr::count() %>%
  filter(grps != "") %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

### Gene counts

```{r table_top_gwas_hot_genes}
gwas_hot %>%
  dplyr::select(gene) %>%
  group_by(gene) %>%
  filter(gene != "") %>%
  dplyr::count() %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```


## CTmin

### All hits (pval cutoff 10^-5)

```{r table_top_gwas_cold_all}
gwas_cold %>%
  dplyr::select(ID, gene, grps, AvgEff, AvgMixedPval) %>%
  arrange(AvgMixedPval) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

### Group count

```{r table_top_gwas_cold_grps}
gwas_cold %>%
  group_by(grps) %>%
  filter(grps != "") %>%
  dplyr::count() %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

### Gene counts

```{r table_top_gwas_cold_genes}
gwas_cold %>%
  dplyr::select(gene) %>%
  group_by(gene) %>%
  filter(gene != "") %>%
  dplyr::count() %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```