---
title: Genome wide association study and differential gene expression of _Drosophila melanogaster_ under heat and cold stress
author: "TS O'Leary"
output:
  rmarkdown::html_document:
    theme: lumen
    number_sections: true
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
# don't show cose, warnings or messages
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

# load these required packages
require(tidyverse)
require(ggpubr)
require(AnnotationDbi)
require(org.Dm.eg.db)
require(ggcorrplot)
require(seqsetvis)
require(ssvRecipes)
require(knitr)
require(kableExtra)

# load the differentially expressed genes results
setwd(here::here("cahan/results"))
res_cold <- read.csv(list.files(pattern = "cold_results"))
res_hot <- read.csv(list.files(pattern = "hot_results"))
deg <- full_join(res_cold, res_hot, by = "gene", suffix = c(".cold", ".hot"))

# load the gwas results
setwd(here::here("DRGP_GWAS"))
gwas_cold <- read.table("CTmin/gwas.top.annot", header = TRUE)
gwas_hot <- read.table("CTmax/gwas.top.annot", header = TRUE)

# filter the gwas so there is only the pvalues that reach the cut off level ----
pval_cut <- 0.00005

gwas_cold <- gwas_cold %>%
  filter(AvgMixedPval < pval_cut)
gwas_hot <- gwas_hot %>%
  filter(AvgMixedPval < pval_cut)

# combine the cold gwas and deg ------------------------------------------------
# use the GeneAnnotation to get the gene symbol for the gwas
annot_split_cold <- str_split(gwas_cold$GeneAnnotation, "\\|")
# the second item is the gene symbol
gwas_cold$gene <- sapply(annot_split_cold, "[[", 2)
# the third item is where the snp is located, intron, syn coding etc.
gwas_cold$grps <- sapply(annot_split_cold, "[[", 3)
# join the two data frames together
comb_cold <- full_join(res_cold, gwas_cold, by = "gene")

# combine the hot gwas and deg -------------------------------------------------
# use the GeneAnnotation to get the gene symbol for the gwas
annot_split_hot <- str_split(gwas_hot$GeneAnnotation, "\\|")
# the second item is the gene symbol
gwas_hot$gene <- sapply(annot_split_hot, "[[", 2)
# the third item is where the snp is located, intron, syn coding etc.
gwas_hot$grps <- sapply(annot_split_hot, "[[", 3)
# join the two data frames together
comb_hot <- full_join(res_hot, gwas_hot, by = "gene")


# create a data.frame with all the info for the plotting later -----------------
temp <- left_join(deg, gwas_cold, by = "gene", suffix = c("", ".cold"))
total <- full_join(temp, gwas_hot, by = "gene", suffix = c(".cold", ".hot"))

deg_grouped <- total %>%
  mutate(cold_genes = case_when(padj.cold < 0.01 ~ "sig",
                                padj.cold >= 0.01 ~ "ns"),
         hot_genes = case_when(padj.hot < 0.01 ~ "sig",
                               padj.hot >= 0.01 ~ "ns"),
         cold_fc = case_when(log2FoldChange.cold < 0 ~ "down",
                             log2FoldChange.cold > 0 ~ "up"),
         hot_fc = case_when(log2FoldChange.hot < 0 ~ "down",
                            log2FoldChange.hot > 0 ~ "up")) %>%
  filter(!is.na(cold_genes) & !is.na(hot_genes) &
           !is.na(cold_fc) & !is.na(hot_fc)) %>%
  mutate(col = paste(cold_genes, hot_genes, cold_fc, hot_fc)) %>%
  mutate(group = case_when(col == "sig sig up up" |
                             col == "sig sig down down" ~ "Shared",
                           col == "sig ns up up" |
                             col == "ns sig up up" |
                             col == "sig ns down down" |
                             col == "ns sig down down" ~ "Sig1 Shared",
                           col == "ns ns down down" |
                             col == "ns ns down up" |
                             col == "ns ns up down" |
                             col == "ns ns up up" ~ "NS",
                           col == "sig sig down up" |
                             col == "sig sig up down" ~ "Unique",
                           col == "ns sig up down" |
                             col == "ns sig down up" |
                             col == "sig ns down up" |
                             col == "sig ns up down" ~ "Sig1 Unique")) %>%
  add_count(group) %>% 
  mutate(groupn = paste0(group, ': ', n)) %>%
  arrange(groupn)



# colors for the shared unique scatter plot            
clrs <- c("darkgrey", "goldenrod4", "goldenrod1", "coral", "coral4")

```

# MA plots {.tabset .tabset-fade .tabset-pills}

## Fold-change cutoff 2

**Cold v Ctrl MA plot**

```{r ma_plot_cold}
ggmaplot(res_cold, main = expression("4°C" %->% "25°C"),
         fdr = 0.05, fc = 2, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_cold$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot}
ggmaplot(res_hot, main = expression("37°C" %->% "25°C"),
         fdr = 0.05, fc = 2, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_hot$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

## No fold-change cutoff

**Cold v Ctrl MA plot**

```{r ma_plot_cold_no_fc}
ggmaplot(res_cold, main = expression("4°C" %->% "25°C"),
         fdr = 0.05, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_cold$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_no_fc}
ggmaplot(res_hot, main = expression("37°C" %->% "25°C"),
         fdr = 0.05, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_hot$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

## No fold-change cutoff (fdr < 0.01)

**Cold v Ctrl MA plot**

```{r ma_plot_cold_no_fcP}
ggmaplot(res_cold, main = expression("4°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_cold$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

**Hot v Ctrl MA plot**

```{r ma_plot_hot_no_fcP}
ggmaplot(res_hot, main = expression("37°C" %->% "25°C"),
         fdr = 0.01, fc = 0, size = 1,
         palette = c("#B31B21", "#1465AC", "#A6A6A680"),
         genenames = as.vector(res_hot$gene),
         legend = "top", top = 20,
         select.top.method = "padj",
         font.label = c("bold", 11), label.rectangle = TRUE,
         font.legend = "bold",
         font.main = "bold",
         ggtheme = ggplot2::theme_minimal()) 
```

# Top 50 differentially expressed genes by padj

## Heat shock

```{r table_top_DEG_hot}
deg_grouped %>%
  distinct(gene, .keep_all = TRUE) %>%
  group_by(hot_fc) %>%
  top_n(n = -50, wt = padj.hot) %>%
  dplyr::select(gene, hot_fc, log2FoldChange.hot, padj.hot) %>%
  arrange(hot_fc, padj.hot) %>%
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 4)))) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

## Cold shock

```{r table_top_DEG_cold}
deg_grouped %>%
  group_by(cold_fc) %>%
  distinct(gene, .keep_all = TRUE) %>%
  top_n(n = -50, wt = padj.cold) %>%
  dplyr::select(gene, cold_fc, log2FoldChange.cold, padj.cold) %>%
  arrange(cold_fc, padj.cold) %>%
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 4)))) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
  
# kable(full_join(deg_cold_table, deg_hot_table, by = "gene")) %>%
#   kable_styling() %>%
#   scroll_box(width = "100%", height = "500px")

```

# Shared and unique - LFC scatter {.tabset .tabset-fade .tabset-pills}

## No labels

```{r shared_unique}
ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  theme(legend.position = "top")
```

## GWAS hits labeled

```{r shared_unique_gwas_lab}
ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, !is.na(ID.cold) | 
                                            !is.na(ID.hot)) %>%
                              distinct(gene, .keep_all = TRUE),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With unique labeled

```{r shared_unique_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Unique") %>%
                              top_n(20, wt = baseMean.cold),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With shared labeled

```{r shared_shared_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Shared") %>%
                              top_n(20, wt = baseMean.cold),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```

## With shared labeled (padj.hot)

```{r shared_shared_padj.hot_labels}
set.seed(1)

ggplot(deg_grouped, 
       aes(x = log2FoldChange.cold, y = log2FoldChange.hot, color = groupn)) +
  geom_point() + 
  scale_color_manual(values = clrs) +
  geom_hline(size = 1, yintercept = 0, color = "black") +
  geom_vline(size = 1, xintercept = 0, color = "black") +
  ylab("Hot vs Ctrl (L2FC)") +
  xlab("Cold vs Ctrl (L2FC)") + 
  theme_classic() + 
  labs(color = "") + 
  ggrepel::geom_label_repel(data = filter(deg_grouped, group == "Shared") %>%
                              top_n(20, wt = padj.hot),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  theme(legend.position = "top")
```


## With density plots

The ultimate goal is to have the density reflect the location of the GWAS hits for CTmin & CTmax

To be fixed soon...


# GWAS

## CTmax

### All hits (pval cutoff 10^-5)

```{r table_top_gwas_hot_all}
gwas_hot %>%
  dplyr::select(ID, gene, grps, AvgEff, AvgMixedPval) %>%
  arrange(AvgMixedPval) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

### Group count

```{r table_top_gwas_hot_grps}
gwas_hot %>%
  group_by(grps) %>%
  count() %>%
  filter(grps != "") %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

### Gene count (>2)

```{r table_top_gwas_hot_genes}
gwas_hot %>%
  dplyr::select(gene) %>%
  group_by(gene) %>%
  filter(gene != "") %>%
  count() %>%
  dplyr::filter(n > 1) %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

## CTmin

### All hits (pval cutoff 10^-5)

```{r table_top_gwas_cold_all}
gwas_cold %>%
  dplyr::select(ID, gene, grps, AvgEff, AvgMixedPval) %>%
  arrange(AvgMixedPval) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "500px")
```

### Group count

```{r table_top_gwas_cold_grps}
gwas_cold %>%
  group_by(grps) %>%
  filter(grps != "") %>%
  count() %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```

### Gene count (>2)

```{r table_top_gwas_cold_genes}
gwas_cold %>%
  dplyr::select(gene) %>%
  group_by(gene) %>%
  filter(gene != "") %>%
  count() %>%
  dplyr::filter(n > 1) %>%
  arrange(desc(n)) %>%
  kable() %>%
  kable_styling("striped") %>%
  scroll_box(width = "100%", height = "100%")
```


# DEG & GWAS {.tabset .tabset-fade .tabset-pills}

## Expression Plots

**Cold vs Ctrl**

```{r DEG_GWAS_cold}

# get the median or average for each treatment 
comb_avg <- comb_cold %>% 
  dplyr::select(contains("_"), gene, ID, padj, log2FoldChange) %>%
  pivot_longer(contains("_"), names_to = "group", values_to = "expression") %>%
  mutate(group = str_replace(group, "_[[:digit:]]*$", "")) %>%
  group_by(gene, group, ID, padj, log2FoldChange) %>%
  summarize(expression = median(expression, na.rm = TRUE)) %>%
  filter(expression > 0) %>%
  pivot_wider(names_from = group, values_from = expression)

comb_sort <- comb_avg %>%
  mutate(g = case_when(is.na(ID) & padj < 0.01 ~ "DEG",
                       is.na(ID) & padj > 0.01 ~ "all",
                       is.na(ID) ~ "all",
                       !is.na(ID) & padj < 0.01 ~ "GWAS-DEG",
                       !is.na(ID) & padj > 0.01 ~ "GWAS-NS")) %>%
  distinct(gene, .keep_all = TRUE) %>%
  group_by(g) %>%
  add_count(g) %>% 
  mutate(groupn = paste0(g, ': ', n)) %>%
  arrange(groupn) %>%
  filter(!is.na(g))  %>%
  distinct(gene, .keep_all = TRUE)

# plot the expression of the ctrl and hot/cold with each group labeled
ggplot(comb_sort, aes(x = ctrl, y = hot)) + 
  geom_point(aes(x = ctrl, y = cold, color = groupn), alpha = 0.7) +
  scale_color_manual(values = c("#999999", "#E69F00", "#ff0000", "#000000")) +
  ggrepel::geom_label_repel(data = filter(comb_sort, g == "GWAS-DEG"),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  scale_x_log10() +
  scale_y_log10() + 
  ylab("Cold (log10 expression)") +
  xlab("Control (log10 expression)") +
  theme_classic() +
  theme(legend.title = element_blank())
```

**Hot vs Ctrl**

```{r DEG_GWAS_hot}
# get the median or average for each treatment 
comb_avg <- comb_hot %>% 
  dplyr::select(contains("_"), gene, ID, padj, log2FoldChange) %>%
  pivot_longer(contains("_"), names_to = "group", values_to = "expression") %>%
  mutate(group = str_replace(group, "_[[:digit:]]*$", "")) %>%
  group_by(gene, group, ID, padj, log2FoldChange) %>%
  summarize(expression = median(expression, na.rm = TRUE)) %>%
  filter(expression > 0) %>%
  pivot_wider(names_from = group, values_from = expression) 

comb_sort <- comb_avg %>%
  mutate(g = case_when(is.na(ID) & padj < 0.01 ~ "DEG",
                       is.na(ID) & padj > 0.01 ~ "all",
                       is.na(ID) ~ "all",
                       !is.na(ID) & padj < 0.01 ~ "GWAS-DEG",
                       !is.na(ID) & padj > 0.01 ~ "GWAS-NS"))  %>%
  distinct(gene, .keep_all = TRUE) %>%
  group_by(g) %>%
  add_count(g) %>% 
  mutate(groupn = paste0(g, ': ', n)) %>%
  arrange(groupn) %>%
  filter(!is.na(g))

# plot the expression of the ctrl and hot/cold with each group labeled
ggplot(comb_sort, aes(x = ctrl, y = hot)) + 
  geom_point(aes(x = ctrl, y = cold, color = groupn), alpha = 0.7) +
  scale_color_manual(values = c("#999999", "#E69F00", "#ff0000", "#000000")) +
  ggrepel::geom_label_repel(data = filter(comb_sort, g == "GWAS-DEG"),
                            aes(label = gene),
                            fill = "white",
                            color = 'black',
                            size = 3.5,
                            show.legend = FALSE) +
  scale_x_log10() +
  scale_y_log10() + 
  ylab("Hot (log10 expression)") +
  xlab("Control (log10 expression)") +
  theme_classic() +
  theme(legend.title = element_blank())
```


## CTmax & DEG Corr

```{r GWAS_CTmax_corr}
y <- comb_hot[c(3, 7, 30:31)]

corr <- round(cor(y, use = "na.or.complete"), 1)
ggcorrplot(corr, hc.order = TRUE, outline.col = "white")

comb_hot_filt <- comb_hot %>%
  filter(!is.na(baseMean) & !is.na(ID))

```

```{r GWAS_CTmax_corr_pval_fc}
ggplot(comb_hot_filt, aes(x = log(AvgMixedPval), y = abs(log2FoldChange), color = grps)) +
  geom_point() 
```

```{r GWAS_CTmax_corr_eff_pval}
ggplot(comb_hot_filt, aes(x = abs(AvgEff), y = log(log2FoldChange), color = grps)) +
  geom_point()
```

```{r GWAS_CTmax_corr_F_pval_fc}
ggplot(comb_hot_filt, aes(x = log(FemaleMixedPval), y = abs(log2FoldChange), color = grps)) +
  geom_point() 
```

```{r GWAS_CTmax_corr_F_eff_pval}
ggplot(comb_hot_filt, aes(x = abs(FemaleEff), y = log(log2FoldChange), color = grps)) +
  geom_point()
```


## CTmin & DEG Corr

```{r GWAS_CTmin_corr}
xcorr <- comb_cold[c(3, 7, 30:31)]

comb_cold_filt <- comb_cold %>%
  filter(!is.na(baseMean) & !is.na(ID))


corr <- round(cor(xcorr, use = "na.or.complete"), 1)
ggcorrplot(corr, hc.order = TRUE, outline.col = "white")
```

```{r GWAS_CTmin_corr_pval_fc_mixed}
ggplot(comb_cold_filt, aes(x = log(AvgMixedPval), y = abs(log2FoldChange), color = grps)) +
  geom_point() 
```

```{r GWAS_CTmin_corr_eff_pvalP}
ggplot(comb_cold_filt, aes(x = abs(AvgEff), y = log(log2FoldChange), color = grps)) +
  geom_point()
```

```{r GWAS_CTmin_corr_F_pval_fc}
ggplot(comb_cold_filt, aes(x = log(FemaleMixedPval), y = abs(log2FoldChange), color = grps)) +
  geom_point() 
```

```{r GWAS_CTmin_corr_F_eff_pval}
ggplot(comb_cold_filt, aes(x = abs(FemaleEff), y = log(log2FoldChange), color = grps)) +
  geom_point()
```